rust/worker/src/storage/s3.rs

```diff
@@ -161,13 +167,88 @@ impl S3Storage {
+        let file_size = tokio::fs::metadata(path)
+            .await
+            .map_err(|err| S3PutError::S3PutError(err.to_string()))?
+            .len();
+        let mut chunk_count = (file_size / MULTIPART_UPLOAD_CHUNK_SIZE) + 1;
+        let mut size_of_last_chunk = file_size % MULTIPART_UPLOAD_CHUNK_SIZE;
+        if size_of_last_chunk == 0 {
+            size_of_last_chunk = MULTIPART_UPLOAD_CHUNK_SIZE;
+            chunk_count -= 1;
+        }
+
+        let upload_id = match self
+            .client
+            .create_multipart_upload()
+            .bucket(&self.bucket)
+            .key(key)
+            .send()
+            .await
+            .map_err(|err| S3PutError::S3PutError(err.to_string()))?
+            .upload_id
+        {
+            Some(upload_id) => upload_id,
+            None => {
+                return Err(S3PutError::S3PutError(
+                    "Multipart upload creation response missing upload ID".to_string(),
+                ));
             }
+        };
+
+        let mut upload_parts = Vec::new();
+        for chunk_index in 0..chunk_count {
+            let this_chunk = if chunk_count - 1 == chunk_index {
+                size_of_last_chunk
+            } else {
+                MULTIPART_UPLOAD_CHUNK_SIZE
+            };
+
+            let stream = ByteStream::read_from()
+                .path(path)
+                .offset(chunk_index * MULTIPART_UPLOAD_CHUNK_SIZE)
+                .length(Length::Exact(this_chunk))
+                .build()
+                .await
+                .map_err(|err| S3PutError::S3PutError(err.to_string()))?;
+
+            //Chunk index needs to start at 0, but part numbers start at 1.
+            let part_number = (chunk_index as i32) + 1;
+            let upload_part_res = self
+                .client
+                .upload_part()
+                .key(key)
+                .bucket(&self.bucket)
+                .upload_id(&upload_id)
+                .body(stream)
+                .part_number(part_number)
+                .send()
+                .await
+                .map_err(|err| S3PutError::S3PutError(err.to_string()))?;
+
+            upload_parts.push(
+                CompletedPart::builder()
+                    .e_tag(upload_part_res.e_tag.unwrap_or_default())
+                    .part_number(part_number)
+                    .build(),
+            );
         }
+
+        self.client
+            .complete_multipart_upload()
+            .bucket(&self.bucket)
+            .key(key)
+            .multipart_upload(
+                CompletedMultipartUpload::builder()
+                    .set_parts(Some(upload_parts))
+                    .build(),
+            )
+            .upload_id(&upload_id)
+            .send()
+            .await
+            .map_err(|err| S3PutError::S3PutError(err.to_string()))?;
+
+        Ok(())
     }
 
     async fn put_bytestream(&self, key: &str, bytestream: ByteStream) -> Result<(), S3PutError> {
```

Comment:
I think this also needs to benefit from multipart

Reply:
11f44fd6